core/test/matrix/batch_dense.cpp

```diff
@@ -0,0 +1,454 @@
+        EXPECT_EQ(m->at(0, 1, 1), value_type{2.5});
+        ASSERT_EQ(m->at(0, 1, 2), value_type{3.5});
+        EXPECT_EQ(m->at(1, 0, 0), value_type{1.0});
+        EXPECT_EQ(m->at(1, 0, 1), value_type{2.5});
+        EXPECT_EQ(m->at(1, 0, 2), value_type{3.0});
+        EXPECT_EQ(m->at(1, 1, 0), value_type{1.0});
+        EXPECT_EQ(m->at(1, 1, 1), value_type{2.0});
+        ASSERT_EQ(m->at(1, 1, 2), value_type{3.0});
+    }
+
+    static void assert_empty(gko::matrix::BatchDense<value_type> *m)
+    {
+        ASSERT_EQ(m->get_num_batches(), 0);
+        ASSERT_EQ(m->get_num_batches(), 0);
+        ASSERT_EQ(m->get_num_stored_elements(), 0);
+    }
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::unique_ptr<gko::matrix::BatchDense<value_type>> mtx;
+};
+
+TYPED_TEST_SUITE(BatchDense, gko::test::ValueTypes);
+
+
+TYPED_TEST(BatchDense, CanBeEmpty)
+{
+    auto empty = gko::matrix::BatchDense<TypeParam>::create(this->exec);
+    this->assert_empty(empty.get());
+}
+
+
+TYPED_TEST(BatchDense, ReturnsNullValuesArrayWhenEmpty)
+{
+    auto empty = gko::matrix::BatchDense<TypeParam>::create(this->exec);
+    ASSERT_EQ(empty->get_const_values(), nullptr);
+}
+
+
+TYPED_TEST(BatchDense, CanBeConstructedWithSize)
+{
+    using size_type = gko::size_type;
+    auto m = gko::matrix::BatchDense<TypeParam>::create(
+        this->exec,
+        std::vector<gko::dim<2>>{gko::dim<2>{2, 4}, gko::dim<2>{2, 3}});
+
+    ASSERT_EQ(m->get_num_batches(), 2);
+    ASSERT_EQ(m->get_size().at(0), gko::dim<2>(2, 4));
+    ASSERT_EQ(m->get_size().at(1), gko::dim<2>(2, 3));
+    EXPECT_EQ(m->get_stride().at(0), 4);
+    EXPECT_EQ(m->get_stride().at(1), 3);
+    ASSERT_EQ(m->get_num_stored_elements(), 14);
+    ASSERT_EQ(m->get_num_stored_elements(0), 8);
+    ASSERT_EQ(m->get_num_stored_elements(1), 6);
+}
+
+
+TYPED_TEST(BatchDense, CanBeConstructedWithSizeAndStride)
+{
+    using size_type = gko::size_type;
+    auto m = gko::matrix::BatchDense<TypeParam>::create(
+        this->exec, std::vector<gko::dim<2>>{gko::dim<2>{2, 3}},
+        std::vector<size_type>{4});
+
+    ASSERT_EQ(m->get_size().at(0), gko::dim<2>(2, 3));
+    EXPECT_EQ(m->get_stride().at(0), 4);
+    ASSERT_EQ(m->get_num_stored_elements(), 8);
+}
+
+
+TYPED_TEST(BatchDense, CanBeConstructedFromExistingData)
+{
+    using value_type = typename TestFixture::value_type;
+    using size_type = gko::size_type;
+    // clang-format off
+    value_type data[] = {
+       1.0, 2.0, -1.0,
+       3.0, 4.0, -1.0,
+       3.0, 5.0, 1.0,
+       5.0, 6.0, -3.0};
+    // clang-format on
+
+    auto m = gko::matrix::BatchDense<TypeParam>::create(
+        this->exec,
+        std::vector<gko::dim<2>>{gko::dim<2>{2, 2}, gko::dim<2>{2, 2}},
+        gko::Array<value_type>::view(this->exec, 12, data),
+        std::vector<size_type>{3, 3});
+
+    ASSERT_EQ(m->get_const_values(), data);
+    ASSERT_EQ(m->at(0, 0, 1), value_type{2.0});
+    ASSERT_EQ(m->at(0, 1, 2), value_type{-1.0});
```

Comment:
No, we have the same behaviour for Dense as well, because technically the view has the data, so we should be able to access it ? And also because the stride > number of columns

Reply:
Eg. with the row_major and block_col_major accessors, if the the index exceeds the stated size, it throws. If Dense allows it, I guess it' fine. But IMO it should be regarded as an error that we "protect" the user against (in debug builds). If the user created a 2x2 view, they must want to access only those 4 entries through the view.

Reply:
But the thing is the user is not creating a (2x2) view. Here the user is creating a view of 9 elements, and the matrix size is indeed 2x2, but it has a stride of 3.

Reply:
Well, if the user passed the dimensions of the two matrices as (2,2), even though they themselves are providing the values array, clearly they don't want to access the remaining entries in the original array through this Dense matrix. Essentially, the gko Dense/BatchDense matrix should be oblivious to and should not affect the entries that are outside the user-specified size. Testing this implies we promise the user they'll be able to access the entries beyond their specified size, which would (or should) never be their intent.